From 25275350e76ab580cb2a31d55144a07d3f7927fc Mon Sep 17 00:00:00 2001
From: Chinmay Deshpande <cddeshpa@uci.edu>
Date: Fri, 5 Jul 2024 16:05:25 -0700
Subject: [PATCH 1/3] [MCAD Patch 0] Add identifier field to mca::Instruction

---
 llvm/include/llvm/MCA/Instruction.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/llvm/include/llvm/MCA/Instruction.h b/llvm/include/llvm/MCA/Instruction.h
index e48a70164bec..4fb0a62ac7f8 100644
--- a/llvm/include/llvm/MCA/Instruction.h
+++ b/llvm/include/llvm/MCA/Instruction.h
@@ -643,6 +643,9 @@ class Instruction : public InstructionBase {
   // True if this instruction has been optimized at register renaming stage.
   bool IsEliminated;
 
+  // Client-dependent identifier for this instruction.
+  std::optional<uint64_t> Identifier;
+
 public:
   Instruction(const InstrDesc &D, const unsigned Opcode)
       : InstructionBase(D, Opcode), Stage(IS_INVALID),
@@ -690,6 +693,9 @@ public:
   bool isRetired() const { return Stage == IS_RETIRED; }
   bool isEliminated() const { return IsEliminated; }
 
+  std::optional<uint64_t> getIdentifier() const { return Identifier; }
+  void setIdentifier(uint64_t Id) { Identifier = Id; }
+
   // Forces a transition from state IS_DISPATCHED to state IS_EXECUTED.
   void forceExecuted();
   void setEliminated() { IsEliminated = true; }
-- 
2.39.5 (Apple Git-154)

